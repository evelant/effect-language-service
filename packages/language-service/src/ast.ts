import * as T from "@effect/core/io/Effect"
import * as CH from "@tsplus/stdlib/collections/Chunk"
import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): ts.FileTextChanges[]
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
    }
    export function applyChanges(text: string, changes: readonly ts.TextChange[]): string
  }
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export const TypeScriptApi = Tag<typeof import("typescript/lib/tsserverlibrary")>()
export const LanguageServiceApi = Tag<ts.LanguageService>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoTypeScriptProgramError {
  readonly _tag = "NoTypeScriptProgramError"
}

export function getProgram() {
  return T.serviceWithEffect(LanguageServiceApi, languageService => T.sync(() => languageService.getProgram()))
    .filterOrFail((program): program is ts.Program => !!program, () => new NoTypeScriptProgramError())
}

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return getProgram().map(program => program.getSourceFile(fileName))
    .filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return T.serviceWith(TypeScriptApi, ts => !!(ts.getCombinedModifierFlags(node) & kind))
}

function getChildNodesContainingRange(
  parent: ts.Node,
  textRange: ts.TextRange
) {
  return Do($ => {
    const ts = $(Effect.service(TypeScriptApi))
    let result: Chunk<ts.Node> = Chunk.empty()

    ts.forEachChild(parent, node => {
      if (node.end < textRange.pos) {
        return
      } else if (node.pos > textRange.end) {
        return
      } else {
        result = result.append(node)
      }
    })

    return result
  })
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  sourceFile: ts.SourceFile,
  textRange: ts.TextRange
) {
  return Do(($) => {
    let result: Chunk<ts.Node> = Chunk.empty()

    $(
      Effect.iterate([sourceFile] as ts.Node[], (_) => _.length > 0)(toProcess => {
        const node = toProcess.shift()!
        return getChildNodesContainingRange(node, textRange).flatMap(_ =>
          Effect.sync(() => {
            result = result.concat(_)
            return toProcess.concat(..._.toArray)
          })
        )
      })
    )

    return CH.from(result.reverse)
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}
