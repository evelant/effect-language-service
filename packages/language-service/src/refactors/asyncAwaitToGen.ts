import * as T from "@effect/core/io/Effect"
import * as AST from "@effect/language-service/ast"
import { createRefactor } from "@effect/language-service/refactors/definition"

export default createRefactor({
  name: "effect/asyncAwaitToGen",
  description: "Convert to Effect.gen",
  apply: (sourceFile, textRange) =>
    Do($ => {
      const ts = $(T.service(AST.TypeScriptApi))
      const languageService = $(T.service(AST.LanguageServiceApi))

      const nodes = $(AST.getNodesContainingRange(sourceFile, textRange))
      return nodes.filter(ts.isFunctionDeclaration).filter(node => !!node.body).filter(node =>
        !!(ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Async)
      ).head.map(node =>
        Do($ => {
          const changeTracker = $(T.service(AST.ChangeTrackerApi))
          const typeChecker = languageService.getProgram()?.getTypeChecker()

          function visitor(node: ts.Node) {
            console.log(
              node.getText(sourceFile),
              typeChecker?.typeToString(typeChecker?.getTypeAtLocation(node))
            )
            ts.visitEachChild(node, visitor, ts.nullTransformationContext)

            return node
          }
          ts.visitEachChild(sourceFile, visitor, ts.nullTransformationContext)

          function logErrors(fileName: string) {
            const allDiagnostics = languageService.getCompilerOptionsDiagnostics()
              .concat(languageService.getSyntacticDiagnostics(fileName))
              .concat(languageService.getSemanticDiagnostics(fileName))

            allDiagnostics.forEach(diagnostic => {
              const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
              if (diagnostic.file) {
                const { character, line } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
                console.log(`  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`)
              } else {
                console.log(`  Error: ${message}`)
              }
            })
          }
          logErrors(sourceFile.fileName)

          let currentFlags = ts.getCombinedModifierFlags(node)
          currentFlags &= ~ts.ModifierFlags.Async
          const newDeclaration = ts.factory.createFunctionDeclaration(
            ts.factory.createModifiersFromModifierFlags(currentFlags),
            node.asteriskToken,
            node.name,
            node.typeParameters,
            node.parameters,
            node.type,
            ts.factory.createBlock([
              ts.factory.createReturnStatement(
                ts.factory.createIdentifier("undef")
              )
            ])
          )

          changeTracker.replaceNode(sourceFile, node, newDeclaration)
        })
      )
    })
})
