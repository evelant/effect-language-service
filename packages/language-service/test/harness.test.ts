import * as AST from "@effect/language-service/ast"
import asyncAwaitToGen from "@effect/language-service/refactors/asyncAwaitToGen"
import type { RefactorDefinition } from "@effect/language-service/refactors/definition"
import * as O from "@tsplus/stdlib/data/Maybe"
import * as fs from "fs"
import ts from "typescript/lib/tsserverlibrary"

function createVirtualFile(fileName: string, sourceText: string) {
  const files = { [fileName]: sourceText }

  const languageServiceHost: ts.LanguageServiceHost = {
    getCompilationSettings() {
      return { ...ts.getDefaultCompilerOptions(), strict: true, target: ts.ScriptTarget.ESNext, noEmit: true }
    },
    getScriptFileNames() {
      return [fileName]
    },
    getScriptVersion(_fileName) {
      return ""
    },
    getScriptSnapshot(fileName) {
      if (fileName in files) {
        return ts.ScriptSnapshot.fromString(files[fileName] || "")
      }
      return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName).toString())
    },
    getCurrentDirectory: () => ".",
    getDefaultLibFileName(options) {
      return ts.getDefaultLibFilePath(options)
    },
    fileExists: (fileName) => {
      if (fileName.endsWith(".d.ts")) {
        return fs.existsSync(fileName)
      }
      return !!files[fileName]
    },
    readFile: (fileName) => {
      if (fileName in files) return files[fileName]
      return fs.readFileSync(fileName).toString()
    }
  }

  const languageService = ts.createLanguageService(
    languageServiceHost,
    undefined,
    ts.LanguageServiceMode.Semantic
  )

  const sourceFile_ = languageService.getProgram()?.getSourceFile(fileName)
  const sourceFile = sourceFile_!

  function applyFileTextChanges(edits: readonly ts.FileTextChanges[]): string {
    const newFiles: Record<string, string> = JSON.parse(JSON.stringify(files))
    for (const fileTextChange of edits) {
      forEachTextChange(fileTextChange.textChanges, (edit) => {
        const content = files[fileTextChange.fileName] || ""
        const prefix = content.substring(0, edit.span.start)
        const middle = edit.newText
        const suffix = content.substring(edit.span.start + edit.span.length)
        newFiles[fileTextChange.fileName] = prefix + middle + suffix
      })
    }
    return newFiles[fileName] || ""
  }

  /** Apply each textChange in order, updating future changes to account for the text offset of previous changes. */
  function forEachTextChange(
    changes: readonly ts.TextChange[],
    cb: (change: ts.TextChange) => void
  ): void {
    // Copy this so we don't ruin someone else's copy
    changes = JSON.parse(JSON.stringify(changes))
    for (let i = 0; i < changes.length; i++) {
      const change = changes[i]!
      cb(change)
      const changeDelta = change.newText.length - change.span.length
      for (let j = i + 1; j < changes.length; j++) {
        if (changes[j]!.span.start >= change.span.start) {
          changes[j]!.span.start += changeDelta
        }
      }
    }
  }

  return {
    languageServiceHost,
    languageService,
    applyFileTextChanges,
    sourceFile
  }
}

export function testRefactorOnExample(refactor: RefactorDefinition, fileName: string) {
  const sourceWithMarker = fs.readFileSync(require.resolve(__dirname + "/../../examples/" + fileName))
    .toString("utf8")
  const cursorPosition = sourceWithMarker.indexOf("/* HERE */")
  const textRange = { pos: cursorPosition, end: cursorPosition + 1 }
  const {
    applyFileTextChanges,
    languageService,
    languageServiceHost,
    sourceFile
  } = createVirtualFile(fileName, sourceWithMarker)

  const canApply = refactor
    .apply(sourceFile, textRange)
    .provideService(AST.TypeScriptApi, ts)
    .provideService(AST.LanguageServiceApi, languageService)
    .unsafeRunSync()

  expect(O.isSome(canApply)).toBe(true)
  if (O.isNone(canApply)) return

  const formatContext = ts.formatting.getFormatContext(
    ts.getDefaultFormatCodeSettings("\n"),
    { getNewLine: () => "\n" }
  )
  const edits = ts.textChanges.ChangeTracker.with(
    {
      formatContext,
      host: languageServiceHost,
      preferences: {}
    },
    (changeTracker) =>
      canApply.value
        .provideService(AST.ChangeTrackerApi, changeTracker)
        .unsafeRunSync()
  )

  expect(applyFileTextChanges(edits)).toMatchSnapshot()
}

function testRefactor(name: string, refactor: RefactorDefinition, fileNames: string[]) {
  for (const fileName of fileNames) {
    describe(fileName, () => {
      it(fileName, () => {
        testRefactorOnExample(refactor, fileName)
      })
    })
  }
}

testRefactor("asyncAwaitToGen", asyncAwaitToGen, ["asyncAwaitToGen.ts"])
