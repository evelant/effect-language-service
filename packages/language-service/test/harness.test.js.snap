// Vitest Snapshot v1

exports[`addPipe.ts > addPipe.ts 1`] = `
"import * as T from \\"@effect/core/io/Effect\\"
import { pipe } from \\"@tsplus/stdlib/data/Function\\"

const test = pipe(1, T.succeed, T.map(/* HERE */(a: number) => a * 2))
"
`;

exports[`removePipe.ts > removePipe.ts 1`] = `
"import * as T from \\"@effect/core/io/Effect\\"
import { pipe } from \\"@tsplus/stdlib/data/Function\\"

const test = /* HERE */
T.map(a => a * 1)(T.succeed(1))
"
`;
